### ISSUE TITLE
Investigate if Data URLs should work and if they do or not

### ISSUE THREAD

**Issue by ericblade** [ericblade/quagga2#189]
per https://github.com/serratus/quaggaJS/issues/433

there may be an issue with using Data URLs in browser? should they work?

**Comment by ericblade** [comment]
@danisss9 So, you can confirm that decodeSingle on a data URI works, in which environment, browser or node, or both?

**Comment by danisss9** [comment]
I tested the decodeSingle() method in browser (Chrome 85) using Quagga 1.2.2 version with the following config:

![Config](https://user-images.githubusercontent.com/36733433/93205171-49bfb900-f74f-11ea-8d44-10630133a840.png)

The ImageUrl parameter is a DataUrl jpeg image taken using GetUserMedia() API (to use device's camera) and exported with Canvas toDataUrl() method.

The result object is one of the following:

- Undefined
- Array of objects with only "Box" propriety
![Result1](https://user-images.githubusercontent.com/36733433/93205062-1846ed80-f74f-11ea-91d0-08ae70c2013d.png)
- Array of objects with "Box", "CodeResult" and other proprieties
![Result2](https://user-images.githubusercontent.com/36733433/93343334-48ae8a80-f828-11ea-88ff-22779ccd3d86.png)

**Comment by ericblade** [comment]
Yeah, that seems about right.  Undefined when nothing at all found, box areas when it finds a potential barcode area, but isn't able to decode it, and full result when it figures it all out.  

Glad to know it works in browser, should probably write a testcase for it, and make sure it works in node as well.

**Comment by danisss9** [comment]
I'm trying to use Quagga to detect and crop barcodes from images. When I'm using Quagga with a video stream it works fine and gives the right "Box" coordinates. 

Now I was trying to do the same using decodeSingle() method and using my own html canvas to crop the img:

![image](https://user-images.githubusercontent.com/36733433/94003684-7dc05d00-fd93-11ea-8247-5d651b39de08.png)

And when I try to draw the detected barcode area this happens: 

![image](https://user-images.githubusercontent.com/36733433/94002763-3d141400-fd92-11ea-9513-c7f2639927a3.png)

So I tried to multiply all the coordinates by 2.5 and it seems better now (but this doesn't work everytime):

![image](https://user-images.githubusercontent.com/36733433/94002944-78164780-fd92-11ea-81d9-a83513b75677.png)
![image](https://user-images.githubusercontent.com/36733433/94003557-48b40a80-fd93-11ea-897a-c8d7d0fbc8d7.png)

I did it by getting the min and max of the X and Y coordinates and using them in canvas drawImage():

![image](https://user-images.githubusercontent.com/36733433/94003175-d3e0d080-fd92-11ea-9774-80608be75e59.png)

The image resolution and canvas size are 1920x1080. Could it be that the algorithm used to detect the barcode has a different resolution? If yes what is its resolution?

**Comment by ericblade** [comment]
Worth noting on the topic of this investigation, is that DataURI is used in the sample that uses file input in browser.  Still there absolutely should be a test that definitely proves that this functionality is tested.  And that it works in node as well.

**Comment by ericblade** [comment]
@danisss9 There is a "size" parameter that can be passed to the inputStream object in Quagga config, which according to somewhere in the code indicates the length of the "longest" side of the image.  There may be problems in setting it as high as 1920, although I did do some changes a while back in an attempt to fix that.  I *believe* it assumes 800 if not specified.

A lot of the locator code remains like magic to me, so I'm not entirely certain why it doesn't just pick these things up automatically.

**Comment by danisss9** [comment]
Yes, it is not possible to set the size to 1920.

![image](https://user-images.githubusercontent.com/36733433/94049095-681b5980-fdcc-11ea-9b8f-a6d4b5fa03ee.png)

But multiplying the coordinates with 2.4 (1920 / 800 = 2.4) seems to work.

**Comment by ericblade** [comment]
That's.. interesting.  It does tell me a couple of things possibly, that that size parameter may primarily be used to adjust the reported coordinates, rather than affecting the actual performance... and also, that i don't have any idea why it'd be throwing that error, as a 36864 length Uint8Array is normally createable.  

SO, that must be happening in barcode_locator.js at either line 67 or 70, where it is using the 3-args format of Uint8Array constructor ... in which it calculates offsets inside the image using the patchSize.  So if that's correct -- and this is all just reading the code, i've not done any further debugging -- then reducing the patch size would probably allow for the larger size to work.  It shouldn't blow up, though, and a larger patchsize should be just fine anyway... so there's definitely something bad happening there.

**Comment by danisss9** [comment]
This only seems to happen when using the "src" parameter and not giving a target inputStream, when you give a InputStream it gets the size of the inputStream automatically and adjusts the coordinates.

It seems the "size" parameter affects the size of the canvas used to locate the barcode. So in my example I set it to 1280 and the 
image is 1920x1080. You take the furthest point (1920), divide by the size (1280) and you get the ratio difference. Then you multiply all the coordinates by that ratio to get the exact points: 

![image](https://user-images.githubusercontent.com/36733433/94152507-8f793180-fe73-11ea-9bcd-f27381cc4f28.png)


About the exception it starts throwing when I set the size to about 1400 or higher (I checked for 1920x1080 images). The exception indicates that is the line 67 like you suspected, but I didn't check the cause of it.

Also, I didn't check if the "patchSize" parameter causes this exception because I just scan with all of them xD

![image](https://user-images.githubusercontent.com/36733433/94153137-51304200-fe74-11ea-99fe-5eb1c22ed5db.png)

**Comment by ericblade** [comment]
Yep, so when you increase the size too far, with the patchSize also up (i think), then you end up with one or both of the last 2 parameters in the constructor being called with a number that is out of bounds of the first parameter.   I think it'll need some debugging to see what the intent of that section of code is, and to see how to repair it properly.

**Comment by ericblade** [comment]
@danisss9 If you're still out here :-) Several issues have been fixed very recently that touch upon this.  1) a pretty nasty bug related to box and boxes locations, although that may have been specifically for area on and locate off .. i'm not entirely sure.. so there may be additional bugs to find still.  2) the buffer size problem .. has been completely eradicated.  Any image size and any patch size combination should "work" now (although if you really crank it up, you can really tank the performance for no useful benefit whatsoever) .. which would probably make your "auto try sizes/patches" a more legit option .. 

The questions about size -- I've finally explored .  When size is given, it actually processes the image *with the longest dimension set to the specified size*.  So, if you give it a 800x600 image, but you tell it size: 1920, it'll expand the 800 side to 1920, and the 600 side proportionally.  If you give it a 1920x1080 image, and tell it size: 800, it'll shrink the 1920 side to 800, and reduce the 1080 side proportionally.  

If you don't provide a size, it'll use the original image size.    *MOST* of the test suites seem to pass best with size set to somewhere between 1.25 and 2x the longest dimension of the image, and with halfSample set true.  I do not know if any of this has to do with the actual images themselves, or with how the decoders/readers work, or some of both.  Some of the test suites (particularly the ones with 600x4420 images ... what??) work best with the size set to around 0.5x the longest size, and halfSample set off. 

I'm wondering how difficult it would be to alter the library itself to scan with several sizes set and several patch sizes.

**Comment by ericblade** [comment]
@danisss9 also, copilot at least, is telling me that box/boxes return values are relative to the image processed size, not the original size, and that explains a whole heck of a lot of what i was seeing as bizarre things when i was trying to draw coordinates in my previous app.  seems like i'll need to make a much more robust sample of how to draw box coordinates, than what we currently have, assuming that what copilot says is true... which does make sense from what you were saying.

Copilot has written quite a detailed documentation that is going to need to be checked out on the topic, it *looks* like it's right, at least until it gets to this part about a built-in overlay canvas that now I'm questioning everything I know about this lib ... :D 

https://github.com/ericblade/quagga2/pull/608/commits/70e926747b6295e4471e00a73ea2b83585ee5796

**Comment by danisss9** [comment]
Hello @ericblade,

Thank you for solving this all this years later. I have been using the workaround I posted here for few years in production now and has been working pretty good. I have been updating the library when available and will test this new version to see if i can remove the workaround. Great to see you still improving it 

Best regards!

**Comment by ericblade** [comment]
@danisss9 cheers! i'm hoping to get the app that i use this in back online this weekend or next, and then i'll be fully testing a bunch of changes that have gone in (mostly copilot authored..... with guidance ... hmm...) and publishing a new release then .. i'm pretty sure that the changes prior to the last few days that are published should not create any issues, but the last few days i've been doing larger changes so i haven't published yet
